### PR TITLE
Add ISRU activation function for Issue #2181

### PR DESCRIPTION
**Pull request by prince776**
I have added ISRU (Inverse Square Root Unit) activation function. Issue #2181 . Please suggest any changes required.

### CODE REVIEW

File: src/mlpack/methods/ann/activation_functions/isru_function.hpp [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  {
+    x = arma::pow(y / x, 3);
+  }
+
+}; // class ISRUFunction
+
+} // namespace ann
+} // namespace mlpack
+
+#endif
```
Comment (himanshupathak21061998) [inline src/mlpack/methods/ann/activation_functions/isru_function.hpp]: Please add a line at the end of file

File: src/mlpack/methods/ann/activation_functions/isru_function.hpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+   * @param y Input activations.
+   * @param x The resulting derivatives. Should be the matrix used to calculate activation y 
+   * @param alpha parameter, default value = 0.1
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Deriv(const InputVecType& y, OutputVecType& x, const double alpha = 0.1)
+  {
+    x = arma::pow(y / x, 3);
+  }
+
```
Comment (himanshupathak21061998) [inline src/mlpack/methods/ann/activation_functions/isru_function.hpp]: Please remove this line for style fix

File: src/mlpack/methods/ann/activation_functions/isru_function.hpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  /**
+   * Computes the ISRU function.
+   *
+   * @param x Input data.
+   * @param y The resulting output activation.
+   * @param alpha parameter, default value = 0.1
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Fn(const InputVecType& x, OutputVecType& y, const double alpha = 0.1)
```
Comment (himanshupathak21061998) [inline src/mlpack/methods/ann/activation_functions/isru_function.hpp]: This line contain more than 80 character

File: src/mlpack/methods/ann/activation_functions/isru_function.hpp [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+  
+  /**
+   * Computes the inverse of the ISRU function
+   * 
+   * @param y Input data.
+   * @param x The resulting inverse of the input data
+   * @param alpha parameter, default value = 0.1
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Inv(const InputVecType& y, OutputVecType& x, const double alpha = 0.1)
```
Comment (himanshupathak21061998) [inline src/mlpack/methods/ann/activation_functions/isru_function.hpp]: Line contain More than 80 char

File: src/mlpack/methods/ann/activation_functions/isru_function.hpp [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+  /**
+   * Computes the first derivative of the ISRU function.
+   *
+   * @param y Input activations.
+   * @param x The resulting derivatives. Should be the matrix used to calculate activation y 
+   * @param alpha parameter, default value = 0.1
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Deriv(const InputVecType& y, OutputVecType& x, const double alpha = 0.1)
```
Comment (himanshupathak21061998) [inline src/mlpack/methods/ann/activation_functions/isru_function.hpp]: Line contain more than 80 char Do it like this 
```
static void Derive(const InputVecType& y, 
                   OutputVecTypes& x, 
```

File: src/mlpack/methods/ann/activation_functions/isru_function.hpp [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+   * Computes the ISRU function using a matrix as a input
+   * 
+   * @param x Input data.
+   * @param y The resulting output activation
+   * @param alpha parameter, default value = 0.1
+   */
+  template<typename eT>
+  static void Fn(const arma::Mat<eT>& x, arma::Mat<eT>& y, const double alpha = 0.1)
+  {
+    y = x / (arma::sqrt(1 + alpha * arma::pow(x,2)));
```
Comment (himanshupathak21061998) [inline src/mlpack/methods/ann/activation_functions/isru_function.hpp]: Add space after comma

Comment (prince776) [reply src/mlpack/methods/ann/activation_functions/isru_function.hpp]: Thanks for pointing out all style errors. I have fixed them.

**Comment by kartikdutt18** [comment]
Hi @prince776, there is a test failing in derivatives check of activation function, can you have a [look](https://dev.azure.com/mlpack/mlpack/_build/results?buildId=442&view=logs&j=9bb2c93b-115c-5be3-b0bf-bdc0e9bf3db4&t=83258a16-80f6-5330-e7ec-6b0fc19c1be7&l=31).
Thanks.

File: src/mlpack/methods/ann/activation_functions/isru_function.hpp [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+ #include <mlpack/prereqs.hpp>
+ 
+namespace mlpack{
+namespace ann /** Artifical Neural Network. */{
+  
+/**
+ * The ISRU function, defined by
+ * 
+ * @f{eqnarray*}{
+ * f(x) &=& x/sqrt(1 + alpha*x*x)
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/activation_functions/isru_function.hpp]: add space between operators. And instead of writing x * x you can write it as x^{2}

File: src/mlpack/methods/ann/activation_functions/isru_function.hpp [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+ * f'(x) &=& cube(1/sqrt(1 + alpha*x*x))
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/activation_functions/isru_function.hpp]: same as above.

File: src/mlpack/methods/ann/activation_functions/isru_function.hpp [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+   * @param alpha parameter, default value = 0.1
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Fn(const InputVecType& x, 
+                 OutputVecType& y, 
+                 const double alpha = 0.1)
+  {
+    y.set_size(arma::size(x));
+
+    for (size_t i = 0; i < x.n_elem; i++)
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/activation_functions/isru_function.hpp]: To keep it consistent, use armadillo to represent this function as done above.

File: src/mlpack/tests/activation_functions_test.cpp [hunk 10]
```diff
@@ -655,4 +656,23 @@ BOOST_AUTO_TEST_CASE(MishFunctionTest)
...
+                                         9.53462589e-01 -9.53462589e-01 \
+                                         1.69030851 0.0");
+
+  const arma::colvec desiredDerivatives("0.603681611 0.34728357 0.190069281 \
+                                         3.13868921e-05 0.866784171 0.866784171 \
+                                         0.603681611 1.0");
+
+  CheckActivationCorrect<ISRUFunction>(activationData, desiredActivations);
+  CheckDerivativeCorrect<ISRUFunction>(desiredActivations,
+      desiredDerivatives);
```
Comment (kartikdutt18) [inline src/mlpack/tests/activation_functions_test.cpp]: Could you adjust tabs here as seen in other tests.

File: src/mlpack/methods/ann/activation_functions/isru_function.hpp [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+   * @param y Input activations.
+   * @param x The resulting derivatives. Should be the matrix used to calculate activation y 
+   * @param alpha parameter, default value = 0.1
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Deriv(const InputVecType& y,
+                    OutputVecType& x, 
+                    const double alpha = 0.1)
+  {
+    x = arma::pow(y / x, 3);
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/activation_functions/isru_function.hpp]: you need to calculate x = Inv(y, alpha) here. Personally I would prefer it if you reduced the expression so that you directly assign value rather than calculating inverse.

**Comment by prince776** [comment]
Yes it's the case for value 0. I'll fix it after coming back from college.

File: src/mlpack/methods/ann/activation_functions/isru_function.hpp (hunk 11 above)
Comment (kartikdutt18) [reply src/mlpack/methods/ann/activation_functions/isru_function.hpp]: I believe this is causing Ctest failures.

**Comment by kartikdutt18** [comment]
> Yes it's the case for value 0. I'll fix it after coming back from college.

Yes I have mentioned it in the review above. Thanks for all your work.

File: src/mlpack/methods/ann/activation_functions/isru_function.hpp (hunk 11 above)
Comment (kartikdutt18) [reply src/mlpack/methods/ann/activation_functions/isru_function.hpp]: I think it would be better to add a fatal error in case of division by zero.

Comment (prince776) [reply src/mlpack/methods/ann/activation_functions/isru_function.hpp]: > I think it would be better to add a fatal error in case of division by zero.

While calculating the inverse of this function, the value gets undefined(in case of division by zero) only when y = 1/alpha^{1/2}, while the range of y = f(x) is (-1/alpha^{1/2} , 1/alpha^{1/2}).
So theoretically that case will never be called however maybe due to precision it can occur. Then would it better to just set y  = 1/alpha^{1/2} - epsilon (say 1e-6) if y >= 1/alpha^{1/2} and set y  = -1/alpha^{1/2} + epsilon if                       y <= -1/alpha^{1/2}. Or should a fatal error be thrown(like you suggested)?

Comment (kartikdutt18) [reply src/mlpack/methods/ann/activation_functions/isru_function.hpp]: I think rather than changing definition to subtract 1e-6, lets throw an error if needed.However I think a member would respond better. I wanted to do so because one might only call derivatives of y for any input. I will take a closer look at the math, if that case is unlikely we shall leave it as it is.

Comment (prince776) [reply src/mlpack/methods/ann/activation_functions/isru_function.hpp]: Ok. Thanks for your input. If you feel like it is necessary to detect those edge cases I'll add check to it. After you review it tomorrow please tell me if anything else should be done.

File: src/mlpack/methods/ann/activation_functions/isru_function.hpp [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+{
+ public:
+  /**
+   * Computes the ISRU function
+   * 
+   * @param x Input data.
+   * @param alpha paramter, default value = 0.1
+   * @return f(x)
+   */
+  static double Fn(const double x, const double alpha = 0.1)
```
Comment (zoq) [inline src/mlpack/methods/ann/activation_functions/isru_function.hpp]: We should implement this as a layer, similar to: https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/layer/elu.hpp#L127 because there is no way, a user can use that function with a base layer and adjust the alpha parameter.

File: src/mlpack/methods/ann/activation_functions/isru_function.hpp [hunk 13]
```diff
@@ -0,0 +1,135 @@
+/**
+ * @file isru_function.hpp
+ * @author Prince Gupta
+ *
+ * Definition and implementation of the ISRU function
```
Comment (zoq) [inline src/mlpack/methods/ann/activation_functions/isru_function.hpp]: Maybe we should mention: ISRU (Inverse Square Root Unit), at least I didn't know what ISRU stands for.

**Comment by kartikdutt18** [comment]
Hi @prince776, could you also fix the merge error ( You can fix it online by clicking resolve merge conflict). Also thanks for making the changes.

File: HISTORY.md [hunk 14]
```diff
@@ -27,8 +27,6 @@
 
   * Add Mish activation function (#2158).
 
-  * Better error handling of eigendecompositions and Cholesky decompositions
```
Comment (kartikdutt18) [inline HISTORY.md]: Could you undo this change, I don't think you were supposed to delete that. Add you function below this.

**Comment by prince776** [comment]
> Hi @prince776, could you also fix the merge error ( You can fix it online by clicking resolve merge conflict). Also thanks for making the changes.

I've done it. Thanks for telling me, I didn't know I could fix this just online.

File: src/mlpack/tests/activation_functions_test.cpp [hunk 15]
```diff
@@ -331,6 +332,53 @@ void CheckPReLUGradientCorrect(const arma::colvec input,
...
+/*
+ * Implementation of the ISRU activation function derivative test.
+ * The derivative function is implemented as LeakyReLU layer in the file
+ * leaky_relu_layer.hpp
+ *
+ * @param input Input data used for evaluating the LeakyReLU activation function.
+ * @param target Target data used to evaluate the LeakyReLU activation.
+ */
+void CheckISRUDerivativeCorrect(const arma::colvec input,
+                                     const arma::colvec target)
```
Comment (kartikdutt18) [inline src/mlpack/tests/activation_functions_test.cpp]: spacing issue.

File: src/mlpack/tests/activation_functions_test.cpp [hunk 16]
```diff
@@ -655,4 +703,23 @@ BOOST_AUTO_TEST_CASE(MishFunctionTest)
...
+ */
+BOOST_AUTO_TEST_CASE(ISRUFunctionTest)
+{
+  const arma::colvec desiredActivations("-1.69030851 2.24928623 \
+                                         2.58731809 -3.160704 \
+                                         9.53462589e-01 -9.53462589e-01 \
+                                         1.69030851 0.0");
+
+  const arma::colvec desiredDerivatives
+                      ("0.603681611 0.34728357 0.190069281 \
```
Comment (kartikdutt18) [inline src/mlpack/tests/activation_functions_test.cpp]: Could start the parenthesis in previous line.

File: src/mlpack/tests/activation_functions_test.cpp [hunk 17]
```diff
@@ -306,6 +306,7 @@ void CheckPReLUDerivativeCorrect(const arma::colvec input,
   }
 }
 
+
```
Comment (kartikdutt18) [inline src/mlpack/tests/activation_functions_test.cpp]: Could you remove this extra empty line.

File: src/mlpack/methods/ann/layer/isru.hpp [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+   * 
+   * @param y Input data.
+   * @param x The resulting inverse of the input data
+   */
+  template<typename InputVecType, typename OutputVecType>
+  static void Inv(const InputVecType& y,
+                  OutputVecType& x,
+                  const double alpha = 0.1)
+  {
+    x = y / arma::sqrt(1 - alpha * arma::pow(y, 2));
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/isru.hpp]: Could you also the condition if  y ==1 / alpha^{2} so that it becomes DBL_MAX.

File: src/mlpack/methods/ann/layer/isru.hpp [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+   * @param x The resulting derivatives. Should be the matrix used to calculate activation y 
+   * @param alpha parameter, default value = 0.1
+   */
+  template<typename InputVecType, typename OutputVecType>
+  void Deriv(const InputVecType& y,
+                    OutputVecType& x)
+  {
+    Inv(y, x, alpha);
+    x = arma::pow(y / x, 3);
+    x.replace(arma::datum::nan, 1);
```
[the remark this answers is not in the document]
Comment (kartikdutt18) [reply src/mlpack/methods/ann/layer/isru.hpp]: I still think we should avoid getting nan, incase any compiler has issue with it. You can write it as

`x = (x != 0) * arma::pow(y / x, 3) + (x == 0) * DBL_MAX;`

[the remark this answers is not in the document]
Comment (prince776) [reply src/mlpack/methods/ann/layer/isru.hpp]: Wow nice catch, I totally didn't consider the case when compiler may have problems with evaluating to nan.

File: src/mlpack/methods/ann/layer/isru.hpp (hunk 18 above)
Comment (prince776) [reply src/mlpack/methods/ann/layer/isru.hpp]: I believe you mean, if y == 1 / alpha^{1/2}, because that is the case when division by zero occurs.

Comment (kartikdutt18) [reply src/mlpack/methods/ann/layer/isru.hpp]: yes, to prevent division by zero.

**Comment by kartikdutt18** [comment]
No worries.

File: src/mlpack/methods/ann/layer/isru.hpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (kartikdutt18) [reply src/mlpack/methods/ann/layer/isru.hpp]: I think rather than this you can use for loop to make y(i) = Deriv(x(i))
and include this condition in Deriv condition in case you haven't already done so.

File: HISTORY.md [hunk 20]
```diff
@@ -28,7 +28,8 @@
   * Add Mish activation function (#2158).
 
   * Better error handling of eigendecompositions and Cholesky decompositions
-    (#2088, #1840).
```
Comment (kartikdutt18) [inline HISTORY.md]: Undo this one too.

File: src/mlpack/tests/activation_functions_test.cpp [hunk 21]
```diff
@@ -331,6 +331,53 @@ void CheckPReLUGradientCorrect(const arma::colvec input,
...
+  isru.Forward(std::move(input), std::move(activations));
+  for (size_t i = 0; i < activations.n_elem; i++)
+  {
+    BOOST_REQUIRE_CLOSE(activations.at(i), target.at(i), 1e-3);
+  }
+}
+
+/*
+ * Implementation of the ISRU activation function derivative test.
+ * The derivative function is implemented as LeakyReLU layer in the file
```
Comment (kartikdutt18) [inline src/mlpack/tests/activation_functions_test.cpp]: Could you fix this typo.

File: src/mlpack/tests/activation_functions_test.cpp [hunk 22]
```diff
@@ -331,6 +331,53 @@ void CheckPReLUGradientCorrect(const arma::colvec input,
...
+ * leaky_relu_layer.hpp
+ *
+ * @param input Input data used for evaluating the LeakyReLU activation function.
```
Comment (kartikdutt18) [inline src/mlpack/tests/activation_functions_test.cpp]: Same as above.

File: src/mlpack/tests/activation_functions_test.cpp [hunk 23]
```diff
@@ -331,6 +331,53 @@ void CheckPReLUGradientCorrect(const arma::colvec input,
   BOOST_REQUIRE_CLOSE(gradient(0), target(0), 1e-3);
 }
 
+
+/*
+ * Implementation of the ISRU activation function test. The function is
+ * implemented as ISRU layer in the file isru.hpp
+ *
+ * @param input Input data used for evaluating the LeakyReLU activation function.
```
Comment (kartikdutt18) [inline src/mlpack/tests/activation_functions_test.cpp]: Could you remove this typo. Thanks.

File: src/mlpack/tests/activation_functions_test.cpp [hunk 24]
```diff
@@ -655,4 +702,23 @@ BOOST_AUTO_TEST_CASE(MishFunctionTest)
   CheckDerivativeCorrect<MishFunction>(desiredActivations,
                                         desiredDerivatives);
 }
+
+/**
+ * Basic test of the ISRU function.
+ */
+BOOST_AUTO_TEST_CASE(ISRUFunctionTest)
+{
+  const arma::colvec desiredActivations("-1.69030851 2.24928623 \
```
Comment (kartikdutt18) [inline src/mlpack/tests/activation_functions_test.cpp]: If possible comment how you got these values like calculated using python interpreter or using tensor flow etc.
Not necessary though but might be helpful for anyone who goes through the test cases.

Comment (prince776) [reply src/mlpack/tests/activation_functions_test.cpp]: I manually calculated them using a calculator. I thought they are not much so manual will be faster.(It wasn't)

Comment (kartikdutt18) [reply src/mlpack/tests/activation_functions_test.cpp]: I think you can leave it then, resolve this conversation.

Comment (kartikdutt18) [reply src/mlpack/tests/activation_functions_test.cpp]: Thanks for the fast patches.

File: src/mlpack/methods/ann/layer/isru.hpp [hunk 25]
```diff
@@ -0,0 +1,195 @@
+/**
+ * @file isru.hpp
+ * @author Prince Gupta
+ *
+ * Definition of ISRU (Inverse Square Root Unit) activation function
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/isru.hpp]: Pedantic style issue, missing stop at the end.

File: src/mlpack/methods/ann/layer/isru.hpp [hunk 26]
```diff
@@ -0,0 +1,195 @@
...
+    typename OutputDataType = arma::mat
+>
+class ISRU
+{
+ public:
+  /**
+   * Create ISRU object using specified hyperparameter alpha.
+   * Default (alpha = 0.1). Alpha should be > 0.
+   * 
+   * @param alpha parameter
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/isru.hpp]: Would be great if we could add a more descriptive parameter description here.

File: src/mlpack/methods/ann/layer/isru.hpp [hunk 27]
```diff
@@ -0,0 +1,195 @@
...
+ private:
+  /**
+   * Computes the value of ISRU activation function.
+   *
+   * @param x Input data.
+   * @return f(x) 
+   */
+  double Fn(const double x)
+  {
+    return x / (std::sqrt(1 + alpha*x*x));
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/isru.hpp]: Pedantic style issue, missing space between `*` operation. Also we could use `std::pow` instead but either is fine.

File: src/mlpack/methods/ann/layer/isru.hpp [hunk 28]
```diff
@@ -0,0 +1,195 @@
...
+  template<typename eT>
+  void Fn(const arma::Mat<eT>& x, arma::Mat<eT>& y)
+  {
+    y = x / (arma::sqrt(1 + alpha * arma::pow(x, 2)));
+  }
+
+  /**
+   * Computes the inverse of the ISRU function
+   *
+   * @param y
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/isru.hpp]: See comment above about the description.

File: src/mlpack/methods/ann/layer/isru.hpp [hunk 29]
```diff
@@ -0,0 +1,195 @@
...
+   * @return f^{-1}(y)
+   */
+  double Inv(const double y)
+  {
+    double yEdge = 1 / std::sqrt(alpha);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/isru.hpp]: I guess, it makes sense to compute the edge parameter once.

File: src/mlpack/methods/ann/layer/isru.hpp [hunk 30]
```diff
@@ -0,0 +1,195 @@
...
+    x.set_size(arma::size(y));
+
+    for (size_t i = 0; i < x.n_elem; i++)
+    {
+      x(i) = Inv(y(i));
+    }
+  }
+
+  /**
+   * Computes the first derivate of the ISRU function
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/isru.hpp]: Missing stop at the end.

File: src/mlpack/methods/ann/layer/isru.hpp [hunk 31]
```diff
@@ -0,0 +1,195 @@
...
+   *
+   * @param y Input activation
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/isru.hpp]: See comment above.

File: src/mlpack/methods/ann/layer/isru.hpp [hunk 32]
```diff
@@ -0,0 +1,195 @@
...
+   * @return f'(x) where f(x) = y
+   */
+  double Deriv(const double y)
+  {
+    if (y == 0)
+      return 1;
+    double x = Inv(y);
+    return std::pow(y / x, 3);
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/isru.hpp]: What about we use:` return std::pow(Inv(y) / x, 3);`, no need to create a copy.

File: src/mlpack/methods/ann/layer/isru_impl.hpp [hunk 33]
```diff
@@ -0,0 +1,60 @@
+/**
+ * @file isru.hpp
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/isru_impl.hpp]: The filename isn't right.

File: src/mlpack/methods/ann/layer/isru_impl.hpp [hunk 34]
```diff
@@ -0,0 +1,60 @@
...
+ * @author Prince Gupta
+ *
+ * Definition of ISRU (Inverse Square Root Unit) activation function
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/isru_impl.hpp]: It's the implementation.

File: src/mlpack/tests/activation_functions_test.cpp [hunk 35]
```diff
@@ -331,6 +331,53 @@ void CheckPReLUGradientCorrect(const arma::colvec input,
...
+
+/*
+ * Implementation of the ISRU activation function test. The function is
+ * implemented as ISRU layer in the file isru.hpp
+ *
+ * @param input Input data used for evaluating the ISRU activation function.
+ * @param target Target data used to evaluate the ISRU activation.
+ */
+void CheckISRUActivationCorrect(const arma::colvec input,
+                                     const arma::colvec target)
```
Comment (zoq) [inline src/mlpack/tests/activation_functions_test.cpp]: This should align with the argument above.

File: src/mlpack/methods/ann/layer/isru.hpp [hunk 36]
```diff
@@ -0,0 +1,195 @@
...
+#ifndef MLPACK_METHODS_ANN_LAYER_ISRU_HPP
+#define MLPACK_METHODS_ANN_LAYER_ISRU_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artifical Neural Network. */ {
+
+/**
+ * The ISRU (Inverse Square Root Unit) activation function, defined by
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/isru.hpp]: Do you think we should cite: https://arxiv.org/abs/1710.09967 here, similar to:

https://github.com/mlpack/mlpack/blob/f68fc14a77fbab24d586092ceac2294f62564b18/src/mlpack/methods/ann/layer/highway.hpp#L39-L49

File: src/mlpack/methods/ann/layer/isru.hpp [hunk 37]
```diff
@@ -0,0 +1,195 @@
...
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class ISRU
+{
+ public:
+  /**
+   * Create ISRU object using specified hyperparameter alpha.
+   * Default (alpha = 0.1). Alpha should be > 0.
+   * ISRU is defined as f(x) = x / (1 + alpha * x^{2})^{1/2}.
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/isru.hpp]: I think we can remove the formula here, as it's already in the header.

File: src/mlpack/methods/ann/layer/isru.hpp [hunk 38]
```diff
@@ -0,0 +1,195 @@
...
+   * 
+   * @param alpha hyperparameter used to calculate ISRU function.
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/isru.hpp]: `hyperparameter` should start with a capital letter. What about we are more descriptive about `alpha`, something like: Hyperparameter alpha controls the value to which an ISRLU saturates for negative inputs (default = 0.1).

**Comment by prince776** [comment]
I'll fix them as soon as I get home. Thanks for the review.

**Comment by zoq** [comment]
Thanks looks like there are some comments that have to be addressed before we can merge this in.

File: src/mlpack/methods/ann/layer/isru_impl.hpp [hunk 39]
```diff
@@ -0,0 +1,78 @@
...
+  double yEdge = 1 / std::sqrt(alpha);
+  double x;
+  g.set_size(arma::size(input));
+  for (size_t i = 0; i < input.n_elem; i++)
+  {
+    if (input(i) == 0)
+    {
+      g(i) = 1;
+      continue;
+    }
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/isru_impl.hpp]: Any particular reason why we need to use `continue` or we could go with an if - else if pair.

File: src/mlpack/methods/ann/layer/isru_impl.hpp [hunk 40]
```diff
@@ -0,0 +1,78 @@
...
+    if (input(i) >= yEdge)
+      x = DBL_MAX;
+    else if (input(i) <= -yEdge)
+      x =  -DBL_MAX;
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/isru_impl.hpp]: We can combine these I guess using signbit (I think it might look cleaner, totally up to you).

File: src/mlpack/methods/ann/layer_names.hpp [hunk 41]
```diff
@@ -372,6 +372,17 @@ class LayerNameVisitor : public boost::static_visitor<std::string>
     return "weightnorm";
   }
 
+  /*
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer_names.hpp]: I think this should be` /**`

Comment (kartikdutt18) [reply src/mlpack/methods/ann/layer_names.hpp]: The whole file has incorrect comment style (You can fix them here and once it gets merged it will be fixed for all). I have made the same changes in my PR.

File: src/mlpack/methods/ann/layer/isru_impl.hpp [hunk 42]
```diff
@@ -0,0 +1,78 @@
...
+#define MLPACK_METHODS_ANN_LAYER_ISRU_IMPL_HPP
+
+// In case it hasn't yet been included.
+#include "isru.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+// This constructor is called for ISRU activation function.
+// 'alpha' is a hyperparameter
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/isru_impl.hpp]: Missing stop at the end.

File: src/mlpack/methods/ann/layer/isru.hpp [hunk 43]
```diff
@@ -0,0 +1,120 @@
...
+  void serialize(Archive& ar, const unsigned int /* version */);
+
+ private:
+  //! Locally-stored delta object.
+  OutputDataType delta;
+
+  //! Locally-stored output parameter object.
+  OutputDataType outputParameter;
+
+  //! ISRU hyperparameter
```
Comment (kartikdutt18) [inline src/mlpack/methods/ann/layer/isru.hpp]: Missing stop at the end.

File: src/mlpack/methods/ann/layer/isru_impl.hpp (hunk 39 above)
Comment (prince776) [reply src/mlpack/methods/ann/layer/isru_impl.hpp]: `if.. else` can work too, I thought it would look cleaner with `continue`. Also if not used then  g(i) will become `nan` due to 0/0 whenever `input(i) == 0`.

Comment (prince776) [reply src/mlpack/methods/ann/layer/isru_impl.hpp]: `if.. else` can work too, I thought it would look cleaner with `continue`. Also if not used then  g(i) will become `nan` due to 0/0 whenever `input(i) == 0`.

File: src/mlpack/methods/ann/layer/isru_impl.hpp (hunk 40 above)
Comment (prince776) [reply src/mlpack/methods/ann/layer/isru_impl.hpp]: Ok, I can change t to use signbit.

File: src/mlpack/methods/ann/layer/isru_impl.hpp (hunk 39 above)
Comment (kartikdutt18) [reply src/mlpack/methods/ann/layer/isru_impl.hpp]: Agreed that the condition is necessary, I think without `continue` it would also be a single line condition like the rest of them.

File: src/mlpack/methods/ann/layer_names.hpp (hunk 41 above)
Comment (prince776) [reply src/mlpack/methods/ann/layer_names.hpp]: I can fix that in #2243 , since it is about fixing `layer_names.hpp`. It also didn't have header guard.

**Comment by rcurtin** [comment]
This code touches the ANN code, and there was recently a big refactoring of the ANN code in #2259, so be sure to merge the master branch into your branch here to make sure that nothing will fail if this PR is merged. :+1: (I'm pasting this message into all possibly relevant PRs.)

**Comment by prince776** [comment]
HI @zoq, @rcurtin . I've been wanting to get this PR finished for some time now. Please review it when you get a chance. I know there are a lot of PRs and issues to deal with, sorry if it's not a good time to remind.

**Comment by zoq** [comment]
@prince776 Sorry for the slow response, looks like there are some open comments that we should address before we merge this in, let me know if you need any help or if I should clarify anything.

**Comment by prince776** [comment]
> @prince776 Sorry for the slow response, looks like there are some open comments that we should address before we merge this in, let me know if you need any help or if I should clarify anything.

I had made all changes as per the reviews. I can't find the ones you're referring to.
